{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "import glob\n",
    "import tabix\n",
    "import csv\n",
    "import numpy as np\n",
    "\n",
    "from itertools import groupby\n",
    "from operator import itemgetter\n",
    "\n",
    "import pyranges as pr\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "def map_chromosome(chrom):\n",
    "\n",
    "\n",
    "    chrom_dict = {\n",
    "        '1':'NC_000001.10',\n",
    "        '2':'NC_000002.11',\n",
    "        '3':'NC_000003.11',\n",
    "        '4':'NC_000004.11',\n",
    "        '5':'NC_000005.9',\n",
    "        '6':'NC_000006.11',\n",
    "        '7':'NC_000007.13',\n",
    "        '8':'NC_000008.10',\n",
    "        '9':'NC_000009.11',\n",
    "        '10':'NC_000010.10',\n",
    "        '11':'NC_000011.9',\n",
    "        '12':'NC_000012.11',\n",
    "        '13':'NC_000013.10',\n",
    "        '14':'NC_000014.8',\n",
    "        '15':'NC_000015.9',\n",
    "        '16':'NC_000016.9',\n",
    "        '17':'NC_000017.10',\n",
    "        '18':'NC_000018.9',\n",
    "        '19':'NC_000019.9',\n",
    "        '20':'NC_000020.10',\n",
    "        '21':'NC_000021.8',\n",
    "        '22':'NC_000022.10',\n",
    "        'X':'NC_000023.10',\n",
    "        'Y':'NC_000024.9',\n",
    "        'MT':'NC_012920.1'\n",
    "    }\n",
    "\n",
    "    return chrom_dict[chrom]\n",
    "\n",
    "\n",
    "\n",
    "def get_overlapping_genome_features(chrom, pos, end, annotations_pr):\n",
    "\n",
    "    chrom = map_chromosome(chrom)\n",
    "\n",
    "    variant = {'Chromosome': [chrom], 'Start': [pos], 'End': [end]}\n",
    "\n",
    "    variant_pr = pr.from_dict(variant)\n",
    "\n",
    "    overlapping_genome_features = variant_pr.join(annotations_pr).as_df()\n",
    "    \n",
    "    return overlapping_genome_features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_genome_features(overlapping_genome_features):\n",
    "\n",
    "    gene_features = {'gene': None, 'pseudogene': None,'miRNA': None,'tRNA': None  }  \n",
    "    transcript_features = {'transcript': None, 'mRNA': None, 'lnc_RNA': None,'primary_transcript': None, 'snoRNA': None, 'antisense_RNA': None, 'snRNA': None, 'guide_RNA': None, 'C_gene_segment': None, 'V_gene_segment': None, 'J_gene_segment': None, 'D_gene_segment': None, 'scRNA': None }\n",
    "\n",
    "    genes = {}\n",
    "    transcripts = {}\n",
    "    transcript_exons = {}\n",
    "\n",
    "    for row in overlapping_genome_features.itertuples():\n",
    "\n",
    "        if row.Feature in gene_features:\n",
    "\n",
    "            if row.Feature == 'miRNA':\n",
    "\n",
    "                biotype = 'miRNA'\n",
    "\n",
    "            elif row.Feature == 'tRNA':\n",
    "\n",
    "                biotype = 'tRNA'\n",
    "\n",
    "            else:\n",
    "\n",
    "                biotype = row.gene_biotype\n",
    "\n",
    "            genes[row.ID] = {'name': row.Name, 'start': row.Start_b, 'end': row.End_b, 'biotype': biotype}\n",
    "\n",
    "\n",
    "        if row.Feature in transcript_features:\n",
    "\n",
    "\n",
    "            if pd.isnull(row.Name):\n",
    "\n",
    "                name = row.ID\n",
    "\n",
    "            else:\n",
    "\n",
    "                name = row.Name\n",
    "\n",
    "            transcripts[row.ID] = {'name': row.Name, 'start': row.Start_b, 'end': row.End_b, 'gene': row.Parent}\n",
    "\n",
    "\n",
    "    for row in overlapping_genome_features.itertuples():\n",
    "\n",
    "        if row.Feature == 'exon':\n",
    "\n",
    "            parent = row.Parent\n",
    "            start = row.Start_b\n",
    "            end = row.End_b\n",
    "\n",
    "            if parent in transcripts:\n",
    "\n",
    "                exon_n = row.ID.split('-')[-1]\n",
    "\n",
    "                if row.transcript_id not in transcript_exons:\n",
    "\n",
    "                    gene_id = genes[transcripts[parent]['gene']]\n",
    "\n",
    "                    transcript_exons[row.transcript_id] = {'exons': [[start, end, exon_n]], 'gene':gene_id }\n",
    "\n",
    "                else:\n",
    "\n",
    "                    transcript_exons[row.transcript_id]['exons'].append([start, end, exon_n])\n",
    "\n",
    "            elif parent in genes:\n",
    "\n",
    "                if parent not in transcript_exons:\n",
    "\n",
    "                    gene_id = genes[parent]\n",
    "\n",
    "                    transcript_exons[parent] = {'exons': [[start, end, 'na']], 'gene':gene_id }\n",
    "\n",
    "                else:\n",
    "\n",
    "                    transcript_exons[parent]['exons'].append([start, end, 'na'])  \n",
    "\n",
    "\n",
    "            else:\n",
    "\n",
    "                print('oh no')\n",
    "                print(parent, start, end)\n",
    "\n",
    "    parsed_overlapping_genome_features = {'transcripts': transcripts, 'genes': genes, 'transcript_exons': transcript_exons}\n",
    "    \n",
    "    return parsed_overlapping_genome_features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "files_to_merge = glob.glob('results/coverage/*_region_coverage_data.csv')\n",
    "\n",
    "master_df = pd.DataFrame()\n",
    "\n",
    "for file in files_to_merge:\n",
    "\n",
    "    df  = pd.read_csv(file)\n",
    "\n",
    "    master_df = master_df.append(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "master_df.head()\n",
    "\n",
    "\n",
    "master_df = master_df.head(1000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "annotations = pr.read_gff3('../../sv_consequences/annotations/GRCh37_latest_genomic.gff.gz')\n",
    "\n",
    "region_annotation_dict = {}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "def annotate_rows(df, annotations):\n",
    "    \n",
    "    chrom = df['chrom']\n",
    "    start = df['start']\n",
    "    end = df['end']\n",
    "    \n",
    "    region_key = f'{chrom}:{start}-{end}'\n",
    "        \n",
    "    if region_key in region_annotation_dict:\n",
    "        \n",
    "        return region_annotation_dict[region_key]\n",
    "    \n",
    "    else:\n",
    "        \n",
    "        feature_df = get_overlapping_genome_features(chrom, int(start), int(end), annotations)\n",
    "        \n",
    "        parsed_features =  parse_genome_features(feature_df)\n",
    "        \n",
    "        region_annotation_dict[region_key] = parsed_features\n",
    "        \n",
    "        return parsed_features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "master_df['annotations'] = master_df.apply(annotate_rows, axis=1,args=(annotations,))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "master_df.to_csv('master.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [],
   "source": [
    "new_rows = []\n",
    "\n",
    "for row in master_df.itertuples():\n",
    "    \n",
    "    anno = row.annotations\n",
    "            \n",
    "    row = [row.mean_coverage, row.region_length, row.error, row.chrom, row.start, row.end, row.gaps, row.pct_gtr_20x, row.pct_gtr_160x, row.sample_id, anno ]\n",
    "    \n",
    "    for ts in anno['transcript_exons']:\n",
    "        \n",
    "        gene = anno['transcript_exons'][ts]['gene']['name']\n",
    "        \n",
    "        exons = []\n",
    "        \n",
    "        for exon in anno['transcript_exons'][ts]['exons']:\n",
    "                        \n",
    "            exons.append(exon[2])       \n",
    "        \n",
    "        new_rows.append(row + [gene] + [exons] + [ts])\n",
    "        \n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [],
   "source": [
    "new_df = pd.DataFrame(new_rows, columns = list(master_df.columns) + ['gene_name', 'exons', 'transcript'] )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mean_coverage</th>\n",
       "      <th>region_length</th>\n",
       "      <th>error</th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>gaps</th>\n",
       "      <th>pct_gtr_20x</th>\n",
       "      <th>pct_gtr_160x</th>\n",
       "      <th>sample_id</th>\n",
       "      <th>annotations</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>exons</th>\n",
       "      <th>transcript</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>31.995851</td>\n",
       "      <td>241</td>\n",
       "      <td>none</td>\n",
       "      <td>1</td>\n",
       "      <td>955532</td>\n",
       "      <td>955773</td>\n",
       "      <td>{20: [[955532, 955617]], 160: [[955532, 955772]]}</td>\n",
       "      <td>0.643154</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>191010_D00501_0366_BH5JWHBCX3_18M01315</td>\n",
       "      <td>{'transcripts': {'rna-NM_001305275.2': {'name'...</td>\n",
       "      <td>AGRN</td>\n",
       "      <td>[1]</td>\n",
       "      <td>NM_001305275.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>31.995851</td>\n",
       "      <td>241</td>\n",
       "      <td>none</td>\n",
       "      <td>1</td>\n",
       "      <td>955532</td>\n",
       "      <td>955773</td>\n",
       "      <td>{20: [[955532, 955617]], 160: [[955532, 955772]]}</td>\n",
       "      <td>0.643154</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>191010_D00501_0366_BH5JWHBCX3_18M01315</td>\n",
       "      <td>{'transcripts': {'rna-NM_001305275.2': {'name'...</td>\n",
       "      <td>AGRN</td>\n",
       "      <td>[1]</td>\n",
       "      <td>NM_198576.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>113.784768</td>\n",
       "      <td>302</td>\n",
       "      <td>none</td>\n",
       "      <td>1</td>\n",
       "      <td>957560</td>\n",
       "      <td>957862</td>\n",
       "      <td>{20: [], 160: [[957560, 957674], [957726, 9578...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.168874</td>\n",
       "      <td>191010_D00501_0366_BH5JWHBCX3_18M01315</td>\n",
       "      <td>{'transcripts': {'rna-NM_001305275.2': {'name'...</td>\n",
       "      <td>AGRN</td>\n",
       "      <td>[2]</td>\n",
       "      <td>NM_001305275.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>113.784768</td>\n",
       "      <td>302</td>\n",
       "      <td>none</td>\n",
       "      <td>1</td>\n",
       "      <td>957560</td>\n",
       "      <td>957862</td>\n",
       "      <td>{20: [], 160: [[957560, 957674], [957726, 9578...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.168874</td>\n",
       "      <td>191010_D00501_0366_BH5JWHBCX3_18M01315</td>\n",
       "      <td>{'transcripts': {'rna-NM_001305275.2': {'name'...</td>\n",
       "      <td>AGRN</td>\n",
       "      <td>[2]</td>\n",
       "      <td>NM_198576.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>145.988636</td>\n",
       "      <td>88</td>\n",
       "      <td>none</td>\n",
       "      <td>1</td>\n",
       "      <td>970636</td>\n",
       "      <td>970724</td>\n",
       "      <td>{20: [], 160: [[970636, 970654], [970681, 9707...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.295455</td>\n",
       "      <td>191010_D00501_0366_BH5JWHBCX3_18M01315</td>\n",
       "      <td>{'transcripts': {'rna-NM_001305275.2': {'name'...</td>\n",
       "      <td>AGRN</td>\n",
       "      <td>[3]</td>\n",
       "      <td>NM_001305275.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3152</th>\n",
       "      <td>197.676000</td>\n",
       "      <td>250</td>\n",
       "      <td>none</td>\n",
       "      <td>1</td>\n",
       "      <td>26135050</td>\n",
       "      <td>26135300</td>\n",
       "      <td>{20: [], 160: [[26135050, 26135084], [26135274...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.756000</td>\n",
       "      <td>191010_D00501_0366_BH5JWHBCX3_18M01315</td>\n",
       "      <td>{'transcripts': {'rna-NM_020451.3': {'name': '...</td>\n",
       "      <td>SELENON</td>\n",
       "      <td>[4]</td>\n",
       "      <td>NM_206926.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3153</th>\n",
       "      <td>124.442424</td>\n",
       "      <td>165</td>\n",
       "      <td>none</td>\n",
       "      <td>1</td>\n",
       "      <td>26135496</td>\n",
       "      <td>26135661</td>\n",
       "      <td>{20: [], 160: [[26135496, 26135660]]}</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>191010_D00501_0366_BH5JWHBCX3_18M01315</td>\n",
       "      <td>{'transcripts': {'rna-NM_020451.3': {'name': '...</td>\n",
       "      <td>SELENON</td>\n",
       "      <td>[6]</td>\n",
       "      <td>NM_020451.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3154</th>\n",
       "      <td>124.442424</td>\n",
       "      <td>165</td>\n",
       "      <td>none</td>\n",
       "      <td>1</td>\n",
       "      <td>26135496</td>\n",
       "      <td>26135661</td>\n",
       "      <td>{20: [], 160: [[26135496, 26135660]]}</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>191010_D00501_0366_BH5JWHBCX3_18M01315</td>\n",
       "      <td>{'transcripts': {'rna-NM_020451.3': {'name': '...</td>\n",
       "      <td>SELENON</td>\n",
       "      <td>[5]</td>\n",
       "      <td>NM_206926.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3155</th>\n",
       "      <td>91.865169</td>\n",
       "      <td>178</td>\n",
       "      <td>none</td>\n",
       "      <td>1</td>\n",
       "      <td>26136153</td>\n",
       "      <td>26136331</td>\n",
       "      <td>{20: [], 160: [[26136153, 26136330]]}</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>191010_D00501_0366_BH5JWHBCX3_18M01315</td>\n",
       "      <td>{'transcripts': {'rna-NM_020451.3': {'name': '...</td>\n",
       "      <td>SELENON</td>\n",
       "      <td>[7]</td>\n",
       "      <td>NM_020451.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3156</th>\n",
       "      <td>91.865169</td>\n",
       "      <td>178</td>\n",
       "      <td>none</td>\n",
       "      <td>1</td>\n",
       "      <td>26136153</td>\n",
       "      <td>26136331</td>\n",
       "      <td>{20: [], 160: [[26136153, 26136330]]}</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>191010_D00501_0366_BH5JWHBCX3_18M01315</td>\n",
       "      <td>{'transcripts': {'rna-NM_020451.3': {'name': '...</td>\n",
       "      <td>SELENON</td>\n",
       "      <td>[6]</td>\n",
       "      <td>NM_206926.2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3157 rows Ã— 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      mean_coverage  region_length error chrom     start       end  \\\n",
       "0         31.995851            241  none     1    955532    955773   \n",
       "1         31.995851            241  none     1    955532    955773   \n",
       "2        113.784768            302  none     1    957560    957862   \n",
       "3        113.784768            302  none     1    957560    957862   \n",
       "4        145.988636             88  none     1    970636    970724   \n",
       "...             ...            ...   ...   ...       ...       ...   \n",
       "3152     197.676000            250  none     1  26135050  26135300   \n",
       "3153     124.442424            165  none     1  26135496  26135661   \n",
       "3154     124.442424            165  none     1  26135496  26135661   \n",
       "3155      91.865169            178  none     1  26136153  26136331   \n",
       "3156      91.865169            178  none     1  26136153  26136331   \n",
       "\n",
       "                                                   gaps  pct_gtr_20x  \\\n",
       "0     {20: [[955532, 955617]], 160: [[955532, 955772]]}     0.643154   \n",
       "1     {20: [[955532, 955617]], 160: [[955532, 955772]]}     0.643154   \n",
       "2     {20: [], 160: [[957560, 957674], [957726, 9578...     1.000000   \n",
       "3     {20: [], 160: [[957560, 957674], [957726, 9578...     1.000000   \n",
       "4     {20: [], 160: [[970636, 970654], [970681, 9707...     1.000000   \n",
       "...                                                 ...          ...   \n",
       "3152  {20: [], 160: [[26135050, 26135084], [26135274...     1.000000   \n",
       "3153              {20: [], 160: [[26135496, 26135660]]}     1.000000   \n",
       "3154              {20: [], 160: [[26135496, 26135660]]}     1.000000   \n",
       "3155              {20: [], 160: [[26136153, 26136330]]}     1.000000   \n",
       "3156              {20: [], 160: [[26136153, 26136330]]}     1.000000   \n",
       "\n",
       "      pct_gtr_160x                               sample_id  \\\n",
       "0         0.000000  191010_D00501_0366_BH5JWHBCX3_18M01315   \n",
       "1         0.000000  191010_D00501_0366_BH5JWHBCX3_18M01315   \n",
       "2         0.168874  191010_D00501_0366_BH5JWHBCX3_18M01315   \n",
       "3         0.168874  191010_D00501_0366_BH5JWHBCX3_18M01315   \n",
       "4         0.295455  191010_D00501_0366_BH5JWHBCX3_18M01315   \n",
       "...            ...                                     ...   \n",
       "3152      0.756000  191010_D00501_0366_BH5JWHBCX3_18M01315   \n",
       "3153      0.000000  191010_D00501_0366_BH5JWHBCX3_18M01315   \n",
       "3154      0.000000  191010_D00501_0366_BH5JWHBCX3_18M01315   \n",
       "3155      0.000000  191010_D00501_0366_BH5JWHBCX3_18M01315   \n",
       "3156      0.000000  191010_D00501_0366_BH5JWHBCX3_18M01315   \n",
       "\n",
       "                                            annotations gene_name exons  \\\n",
       "0     {'transcripts': {'rna-NM_001305275.2': {'name'...      AGRN   [1]   \n",
       "1     {'transcripts': {'rna-NM_001305275.2': {'name'...      AGRN   [1]   \n",
       "2     {'transcripts': {'rna-NM_001305275.2': {'name'...      AGRN   [2]   \n",
       "3     {'transcripts': {'rna-NM_001305275.2': {'name'...      AGRN   [2]   \n",
       "4     {'transcripts': {'rna-NM_001305275.2': {'name'...      AGRN   [3]   \n",
       "...                                                 ...       ...   ...   \n",
       "3152  {'transcripts': {'rna-NM_020451.3': {'name': '...   SELENON   [4]   \n",
       "3153  {'transcripts': {'rna-NM_020451.3': {'name': '...   SELENON   [6]   \n",
       "3154  {'transcripts': {'rna-NM_020451.3': {'name': '...   SELENON   [5]   \n",
       "3155  {'transcripts': {'rna-NM_020451.3': {'name': '...   SELENON   [7]   \n",
       "3156  {'transcripts': {'rna-NM_020451.3': {'name': '...   SELENON   [6]   \n",
       "\n",
       "          transcript  \n",
       "0     NM_001305275.2  \n",
       "1        NM_198576.4  \n",
       "2     NM_001305275.2  \n",
       "3        NM_198576.4  \n",
       "4     NM_001305275.2  \n",
       "...              ...  \n",
       "3152     NM_206926.2  \n",
       "3153     NM_020451.3  \n",
       "3154     NM_206926.2  \n",
       "3155     NM_020451.3  \n",
       "3156     NM_206926.2  \n",
       "\n",
       "[3157 rows x 14 columns]"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "new_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mean_coverage</th>\n",
       "      <th>region_length</th>\n",
       "      <th>error</th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>gaps</th>\n",
       "      <th>pct_gtr_20x</th>\n",
       "      <th>pct_gtr_160x</th>\n",
       "      <th>sample_id</th>\n",
       "      <th>annotations</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>exons</th>\n",
       "      <th>transcript</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [mean_coverage, region_length, error, chrom, start, end, gaps, pct_gtr_20x, pct_gtr_160x, sample_id, annotations, gene_name, exons, transcript]\n",
       "Index: []"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "new_df[new_df['gene_name']=='RNU4ATAC']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
